svc.go

```diff
@@ -0,0 +1,181 @@
+package autonat
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	pb "github.com/libp2p/go-libp2p-autonat/pb"
+
+	ggio "github.com/gogo/protobuf/io"
+	libp2p "github.com/libp2p/go-libp2p"
+	host "github.com/libp2p/go-libp2p-host"
+	inet "github.com/libp2p/go-libp2p-net"
+	peer "github.com/libp2p/go-libp2p-peer"
+	pstore "github.com/libp2p/go-libp2p-peerstore"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+const P_CIRCUIT = 290
+
+var AutoNATServiceResetInterval = 1 * time.Minute
+
+type AutoNATService struct {
+	ctx    context.Context
+	dialer host.Host
+	peers  map[peer.ID]struct{}
+	mx     sync.Mutex
+}
+
+func NewAutoNATService(ctx context.Context, h host.Host) (*AutoNATService, error) {
+	dialer, err := libp2p.New(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	as := &AutoNATService{
+		ctx:    ctx,
+		dialer: dialer,
+		peers:  make(map[peer.ID]struct{}),
+	}
+	h.SetStreamHandler(AutoNATProto, as.handleStream)
+
+	go as.resetPeers()
+
+	return as, nil
+}
+
+func (as *AutoNATService) handleStream(s inet.Stream) {
+	defer s.Close()
+
+	pid := s.Conn().RemotePeer()
+	log.Debugf("New stream from %s", pid.Pretty())
+
+	r := ggio.NewDelimitedReader(s, inet.MessageSizeMax)
+	w := ggio.NewDelimitedWriter(s)
+
+	var req pb.Message
+	var res pb.Message
+
+	err := r.ReadMsg(&req)
+	if err != nil {
+		log.Debugf("Error reading message from %s: %s", pid.Pretty(), err.Error())
+		s.Reset()
+		return
+	}
+
+	t := req.GetType()
+	if t != pb.Message_DIAL {
+		log.Debugf("Unexpected message from %s: %s (%d)", pid.Pretty(), t.String(), t)
+		s.Reset()
+		return
+	}
+
+	dr := as.handleDial(pid, req.GetDial().GetPeer())
+	res.Type = pb.Message_DIAL_RESPONSE.Enum()
+	res.DialResponse = dr
+
+	err = w.WriteMsg(&res)
+	if err != nil {
+		log.Debugf("Error writing response to %s: %s", pid.Pretty(), err.Error())
+		s.Reset()
+		return
+	}
+}
+
+func (as *AutoNATService) handleDial(p peer.ID, mpi *pb.Message_PeerInfo) *pb.Message_DialResponse {
+	if mpi == nil {
+		return newDialResponseError(pb.Message_E_BAD_REQUEST, "missing peer info")
+	}
+
+	mpid := mpi.GetId()
+	if mpid != nil {
+		mp, err := peer.IDFromBytes(mpid)
+		if err != nil {
+			return newDialResponseError(pb.Message_E_BAD_REQUEST, "bad peer id")
+		}
+
+		if mp != p {
+			return newDialResponseError(pb.Message_E_BAD_REQUEST, "peer id mismatch")
+		}
+	}
+
+	addrs := make([]ma.Multiaddr, 0)
+	for _, maddr := range mpi.GetAddrs() {
+		addr, err := ma.NewMultiaddrBytes(maddr)
+		if err != nil {
+			log.Debugf("Error parsing multiaddr: %s", err.Error())
+			continue
+		}
+
+		// skip relay addresses
+		_, err = addr.ValueForProtocol(P_CIRCUIT)
```

Comment:
I'd filter for `/ip[4,6]` here

Reply:
this is handled in the address filter; see `isPublicAddr` in `addr.go`.